app/components/styleguides/button.js

```diff
@@ -0,0 +1,82 @@
+import Component from '@ember/component';
+import abstractMethod from 'travis/utils/abstract-method';
+import { computed } from '@ember/object';
+
+const DEFAULT_SHADES = {
+  light: 100,
+  main: 400,
+  dark: 500,
+};
+
+const DEFAULT_COLOR = 'blue';
+const COLORSETS = {
+  disabled: {
```

Comment:
Yeah I do feel it is a bit too complicated. The goal was to help manage button states for different variants. For example, the fill button generally needs a dark shade background on hover and the outline button needs a light shade background on hover. Though we can maybe use opacity or do something else instead.

I like `disable` as its own state, although I'm not sure of the mechanics of using it to set attributes that are defined elsewhere. Depending on what you're trying to do, I think you may need to remove some of the other classes. But maybe `disabled` could be another parameter for the `buildClassesFor` function?

Reply:
I spent some more time thinking about the components configuration approach. Here's what I ended up with. 

First I tried to imagine how would template of a component look so that it was declarative, configurable, flexible and extendible, but still clear and easy to understand. And that's what I drafted (it's somewhat metaphorical):
```hbs
<button 
  class="
    bg-{{this.bgColor}} 
    border-{{this.borderColor}}
    hover:bg-{{this.hoverBgColor}} 
    hover:border-{{this.hoverBorderColor}}
  " 
  {{on 'click' (fn this.handleClick)}}
>
  {{#if this.hasIcon}}
    <Icon @name={{@icon}} size={{this.iconSize}} />
  {{/if}}
  {{#if this.hasLabel}}
    <Typography 
      @variant='button-label' 
      @size={{this.labelSize}} 
      @color={{this.labelColor}} 
      @tag='span'
    >
      {{@label}}
    </Typography>
  {{/if}}
</button>
```
In this template we need some computed properties that derive values from component configuration. So component class could look like this:
```js
export default Component.extend({

  // public interface
  
  get color() {
    return this.args.color || DEFAULT_COLOR;
  },
  
  get size() {
    return this.args.size || DEFAULT_SIZE;
  },

  get disabled() {
    return this.attrs.disabled || false;
  },
  
  onClick() {},
  
  // private
  
  bgColor: computed('color', 'disabled', function () {
    return this.disabled ? BG_COLORS['disabled'] : BG_COLORS[this.color];
  }),
  
  hoverBgColor: computed('color', 'disabled', function () {
    return this.disabled ? HOVER_BG_COLORS['disabled'] : HOVER_BG_COLORS[this.color];
  }),
  
  labelSize: computed('size', function () {
    return LABEL_SIZES[this.size];
  }),
  
  init() {
    this._super(...arguments);
    assert(`Color "${this.color}" is not allowed on this component`, Object.values(COLORS).includes(this.color));
    assert(`Size "${this.size}" is not allowed on this component`, Object.values(SIZES).includes(this.size));
  },
  
  actions: {
  
    handleClick() {
      this.onClick();
    }
  
  }

})
```
All computed values are pretty much static 1-1 map to the tracked public property, so they could be stored in some kind of configuration objects. I put them to plain dictionaries like this:
```js
// public dictionaries

export const COLORS = {
  GREEN: 'green-400',
  BLUE: 'blue-400'
};

export const SIZES = {
  SM: 'sm'
};

// private dictionaries

const BG_COLORS = {
  [COLORS.BLUE]: 'blue-800',
  [COLORS.GREEN]: 'green-800',
  disabled: 'grey-800'
};

const LABEL_SIZES = {
  [SIZES.SM]: 'xs'
};
```
But we could also use some kind of "theme" global config for such values. Though I find it useful to have them on the same file - just for a quick reference.

What do you think? It's not something I insist on, or even recommend, it's just the idea that I ended up with while thinking, most like a suggestion :)

Reply:
I like this a lot! I made some updates to the button component that heavily borrowed from this. I'm leaving `size` out for now since it's not needed yet for the plans page, and also because I believe that this approach can be utilized for it.

I added below the plans boxes an example of a grey invert button that I will need for the next section, to test and demonstrate what the added `invert` state is for. I feel that there are maybe some improvements possible still for the button component, but it's at least in a much better state now.